lit_nlp/client/modules/perturbations_table_module.ts

```diff
@@ -0,0 +1,310 @@
+  static title = 'Perturbations';
+  static numCols = 4;
+  static duplicateForModelComparison = true;
+  static duplicateAsRow = false;
+  static template = (model = '') => {
+    return html`<perturbations-table-module model=${model}></perturbations-table-module>`;
+  };
+
+  static shouldDisplayModule(modelSpecs: ModelsMap, datasetSpec: Spec) {
+    return doesOutputSpecContain(modelSpecs, [
+      'RegressionScore',
+      'MulticlassPreds'
+    ]);
+  }
+
+  static get styles() {
+    return [sharedStyles, styles];
+  }
+
+  private readonly regressionService = app.getService(RegressionService);
+  private readonly classificationService = app.getService(ClassificationService);
+  private readonly deltasService = app.getService(DeltasService);
+
+  /* These constants are for referring to table columns */
+  private readonly ID_COLUMN = 0;
+  private readonly SENTENCE_COLUMN = 1;
+  private readonly DELTA_COLUMN = 5;
+
+  @observable private filterSelected = true;
+  @observable private lastSelectedSourceIndex?: number;
+
+  private onSelect(selectedRowIndices: number[]) {
+    const ids = selectedRowIndices
+                    .map(index => this.appState.currentInputData[index]?.id)
+                    .filter(id => id != null);
+    this.selectionService.selectIds(ids);
+  }
+
+  private onPrimarySelect(index: number) {
+    const id = (index === -1)
+      ? null
+      : this.appState.currentInputData[index]?.id ?? null;
+    this.selectionService.setPrimarySelection(id);
+  }
+
+  render() {
+    const ds = this.appState.generatedDataPoints;
+    if (ds.length === 0) {
+      return html`<div class="info">No counterfactuals created yet.</div>`;
+    }
+
+    /* Consider classification and regression predictions, and fan out by
+     * each (model, outputKey, fieldName)
+     */
+    return this.deltasService.sourcesForModel(this.model).map((source, index) => {
+      return this.renderHeaderAndTable(source, index);
+    });
+
+  }
+
+  private renderHeaderAndTable(source: Source, sourceIndex: number) {
+    const {fieldName} = source;
+
+    const rootClass = classMap({
+      'hidden': (sourceIndex !== (this.lastSelectedSourceIndex ?? 0)),
+      [sourceIndex]: true
+    });
+    const {generationKeys, deltaRows} = this.deltasService.deltaInfoFromSource(source);
+    const filteredDeltaRows = this.filteredDeltaRows(deltaRows);
+    const tableRows = this.formattedTableRows(filteredDeltaRows);
+    const deltaRowsById: DeltaRowsById = {};
+    deltaRows.forEach(deltaRow => deltaRowsById[deltaRow.d.id] = deltaRow);
+    return html`
+      <div class=${rootClass}>
+        ${this.renderHeader(generationKeys.length, tableRows.length, sourceIndex)}
+        ${this.renderTable(source, tableRows, deltaRowsById)}
+      </div>
+    `;
+  }
+
+  private renderHeader(generationsCount: number, rowsCount: number, sourceIndex: number) {
+    const toggleFilterSelected = () => {
+      this.filterSelected = !this.filterSelected;
```

Comment:
what does this toggle do? couldn't tell from running the PR

Reply:
oops, nothing :)  This is a mistake from shuffling branches; this should be checked in the `filteredDeltaRows` method, I'll update.

Reply:
fixed